### PR TITLE
Mark exceed activeDeadlineSeconds job with completions == succeed pod as Complete instead of Failed 

### LINKED ISSUE #46497: Mark ExceedDeadline Job with completions == succeed instead of  Failed?
**this a FEATURE REQUEST**:

Should we mark exceedDeadline Job(completions: 1/parallelism: 1) with 1 Pod succeed as Completed instead of Failed? I think it's a little wired, my expected Pod(completions: 1/parallelism: 1) has succeed, but the Job is failed.

**Kubernetes version** (use `kubectl version`): version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.1"

**Environment**:
- **Cloud provider or hardware configuration**: baremetal
- **OS** (e.g. from /etc/os-release): 
NAME="Ubuntu"
VERSION="16.04.2 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.2 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
- **Kernel** (e.g. `uname -a`):
Linux xs342 4.4.0-72-generic #93-Ubuntu SMP Fri Mar 31 14:07:41 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

**What happened**:
I use the job.yaml as below

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: test-job-deadline
  labels:
    name: test-job-deadline
spec:
  activeDeadlineSeconds: 5
  completions: 1
  template:
    metadata:
      labels:
        name: test-job-deadline
    spec:
      containers:
      - name: test
        image: busybox:1.26.2-glibc
        args:
        - sleep 10 && echo done && exit 0
        command:
        - sh
        - -c
        imagePullPolicy: IfNotPresent
      restartPolicy: Never
```

and got job Failed with 1 Pod succeeded, here's the kubectl get job
```yaml
status:
  conditions:
  - lastProbeTime: 2017-05-26T06:19:46Z
    lastTransitionTime: 2017-05-26T06:19:46Z
    message: Job was active longer than specified deadline
    reason: DeadlineExceeded
    status: "True"
    type: Failed
  startTime: 2017-05-26T06:19:33Z
  succeeded: 1
```

**What you expected to happen**:

Should we consider Job(completions: 1/parallelism: 1) with 1 Pod succeed as Completed even the Job has exceed the activeDeadline?

I spend some time to read the code, It appears because:
the Pod success kick a Update Event, and the JobController syncJob("test-job-deadline") first check the job [pastActiveDeadline](https://github.com/kubernetes/kubernetes/blob/release-1.6/pkg/controller/job/jobcontroller.go#L334) and mark the Job Failed.

If i set 'sleep SOMETIME && echo done && exit 0' the SOMETIME is longer than 2 * min-resync-period, the result was as expected Job Failed with succeed: 0, failed: 1, that's because the Event was kicked by informers Sync.

**How to reproduce it** (as minimally and precisely as possible):

kubectl apply -f job.yaml

### PR DESCRIPTION
**Pull request by weiwei04**
**What this PR does / why we need it**:

Consider a scenario, the jobSync was kick by the last succeed Pod in a Job, jobSync checked the job has past the actibeDeadlineSeconds and mark the Job condition as Failed, but the job .spec.completions == succeed pod count. Will it be better if we mark the Job as succeed instead of Failed?

**Which issue this PR fixes**: fixes #46497

**Special notes for your reviewer**:

I will add some test if we can mark Job as completed for this scenario:-)

### CODE REVIEW

**Comment by janetkuo** [comment]
@k8s-bot ok to test

**Comment by weiwei04** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by weiwei04** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by weiwei04** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by soltysh** [comment]
@weiwei04 sorry, I just got back from my trip, will look into your PR tomorrow.

**Comment by weiwei04** [comment]
@soltysh no problem, thanks!

**Comment by soltysh** [comment]
@weiwei04 the change itself lgtm, but I'd like to have a test case proving your fix to be able to merge it in. Thanks!

**Comment by weiwei04** [comment]
@soltysh  add ut and amend the commit log, please take a look, thanks!

**Comment by weiwei04** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Comment by weiwei04** [comment]
@soltysh add ut and amend the commit log, the test have passed, please take a look thanks!

**Comment by weiwei04** [comment]
@soltysh any modification suggestion？

File: pkg/controller/job/jobcontroller.go [hunk 1]
```diff
@@ -431,7 +431,41 @@ func (jm *JobController) syncJob(key string) error {
...
+				jm.recorder.Event(&job, v1.EventTypeWarning, "TooManySucceededPods", "Too many succeeded pods running after completion count reached")
+			}
+		}
+	}
+
+	if jobComplete {
+		job.Status.Conditions = append(job.Status.Conditions, newCondition(batch.JobComplete, "", ""))
+		now := metav1.Now()
+		job.Status.CompletionTime = &now
+	} else if jobTimeout {
```
Comment (janetkuo) [inline pkg/controller/job/jobcontroller.go]: With this new logic, a job can be both complete and timeout, and such job won't kill its active pods while it should

Comment (weiwei04) [inline pkg/controller/job/jobcontroller.go]: Sorry for the late reply, thanks for point out the case, there's a condition may left active pods: the syncJob kicked by a success pod and there's still have active pods running for this job, the new logic mark the job as complete and left the active pods run to terminate, it's better to delete these active pods to release resources.

**Comment by weiwei04** [comment]
I think it's ok to close this pr, since #48454 would be a better solution.

**Comment by soltysh** [comment]
+1 to what @weiwei04 said before. Closing it.

**Pull request closed without merging**